plugins/osdn/common.go

```diff
@@ -140,64 +138,72 @@ func (oc *OsdnController) validateNetworkConfig(clusterNetwork, serviceNetwork *
+	oldNetwork, err := oc.Registry.GetNetworkInfo()
 	if err != nil {
 		return false, err
 	}
-	if clusterNetworkCIDR != clusterNetwork.String() ||
-		hostSubnetLength != hostBitsPerSubnet ||
-		serviceNetworkCIDR != serviceNetwork.String() {
+
+	if curNetwork.ClusterNetwork.String() != oldNetwork.ClusterNetwork.String() ||
+		curNetwork.HostSubnetLength != oldNetwork.HostSubnetLength ||
+		curNetwork.ServiceNetwork.String() != oldNetwork.ServiceNetwork.String() ||
+		curNetwork.PluginName != oldNetwork.PluginName {
 		return true, nil
 	}
 	return false, nil
 }
 
 func (oc *OsdnController) StartMaster(clusterNetworkCIDR string, clusterBitsPerSubnet uint, serviceNetworkCIDR string) error {
 	// Validate command-line/config parameters
-	hostBitsPerSubnet := int(clusterBitsPerSubnet)
-	clusterNetwork, _, serviceNetwork, err := ValidateClusterNetwork(clusterNetworkCIDR, hostBitsPerSubnet, serviceNetworkCIDR)
+	network, err := ValidateClusterNetwork(clusterNetworkCIDR, int(clusterBitsPerSubnet), serviceNetworkCIDR, oc.pluginName)
 	if err != nil {
 		return err
 	}
 
-	changed, net_err := oc.isClusterNetworkChanged(clusterNetworkCIDR, hostBitsPerSubnet, serviceNetworkCIDR)
+	changed, net_err := oc.isClusterNetworkChanged(network)
 	if changed {
-		if err := oc.validateNetworkConfig(clusterNetwork, serviceNetwork); err != nil {
+		if err := oc.validateNetworkConfig(network); err != nil {
 			return err
 		}
-		if err := oc.Registry.UpdateClusterNetwork(clusterNetwork, hostBitsPerSubnet, serviceNetwork); err != nil {
+		if err := oc.Registry.UpdateClusterNetwork(network); err != nil {
 			return err
 		}
 	} else if net_err != nil {
-		if err := oc.Registry.CreateClusterNetwork(clusterNetwork, hostBitsPerSubnet, serviceNetwork); err != nil {
+		if err := oc.Registry.CreateClusterNetwork(network); err != nil {
 			return err
 		}
 	}
 
-	if err := oc.pluginHooks.PluginStartMaster(clusterNetwork, clusterBitsPerSubnet); err != nil {
+	if err := oc.pluginHooks.PluginStartMaster(network.ClusterNetwork, clusterBitsPerSubnet); err != nil {
 		return fmt.Errorf("Failed to start plugin: %v", err)
 	}
 	return nil
 }
 
 func (oc *OsdnController) StartNode(mtu uint) error {
 	// Assume we are working with IPv4
-	clusterNetwork, err := oc.Registry.GetClusterNetwork()
+	ni, err := oc.Registry.GetNetworkInfo()
 	if err != nil {
-		log.Errorf("Failed to obtain ClusterNetwork: %v", err)
-		return err
+		return fmt.Errorf("Failed to get network information: %v", err)
+	}
+	if ni.PluginName != oc.pluginName {
+		if ni.PluginName == "" {
+			// Do not return error in this case
+			log.Warningf("Either there is network plugin mismatch between OpenShift master(%q) and node(%q) or OpenShift master is running older version where we don't persist plugin name")
+		} else {
+			return fmt.Errorf("Detected network plugin mismatch between OpenShift master(%q) and node(%q)", ni.PluginName, oc.pluginName)
```

Comment:
should we (in another PR) deprecate the node-config networkPluginName argument and just get that value from the master like we do with the rest of the network config?